Feature/notification system/notification settings

Malyuk-A opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [ ] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [ ] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [ ] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [ ] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [ ] Server: I implemented the changes with a good performance and prevented too many database calls
- [ ] Server: I documented the Java code using JavaDoc style.
- [ ] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [ ] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [ ] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [ ] Client: I documented the TypeScript code using JSDoc style.
- [ ] Client: I added multiple screenshots/screencasts of my UI changes
- [ ] Client: I translated all newly inserted strings into English and German.
- [ ] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

### Description
<!-- Describe your changes in detail -->

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to Course Administration
3. ...

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

- Code Review
  - [ ] Review 1
  - [ ] Review 2
- Manual Tests
  - [ ] Test 1
  - [ ] Test 2

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
<!-- * ExerciseService.java: 85% -->
<!-- * programming-exercise.component.ts 95% -->

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

ndricimrr commented on src/main/webapp/app/shared/user-settings/user-settings-prototype/user-settings-prototype.component.ts:
```diff
@@ -0,0 +1,106 @@
+ * Is used as the abstract user-settings "parent" component with all the necessary basic logic for other "child" components to implement/inherit from.
+ */
+@Component({
+    templateUrl: 'user-settings-prototype.component.html',
```
(i) How does this relate to the file: _"user-settings-prototype.component.html"_ . I see that file is commented.
(i) Do we even need the `@Component` directive in this case, if the template is not being used.

Malyuk-A replied:
![image](https://user-images.githubusercontent.com/65814168/132983530-58c05b49-807c-47d9-b45e-ade266911dd0.png)

As far as I know I sadly have to use the @Component decorator and a valid templateUrl, otherwise error occur during compilation, because I am using OnInit and I can only do so if the class has such a decorator.

The reason why the html is commented out is due to the reason that it is not possible to reuse this template in a polymorphic/inheritance kind of way due to the different implementation of the concrete OptionCore and its bidirectional connectivity to the template (e.g. you need to access the option that is only possible due to the ngFor loop but this creates huge problems for reusability. Therefore I commented this template out to provide an examplary template that can be easily copied for new settings.

![image](https://user-images.githubusercontent.com/65814168/132983549-a2dc811d-c439-49c2-a83a-bb11f4d4c6e0.png)

ndricimrr reviewed (commented):
Interesting PR, although a bit large. :) . I left some comments on the code and there are about ~3 'Codacy' comments I saw along the lines, that might need addressing.

Malyuk-A commented on src/main/webapp/app/shared/user-settings/user-settings-guide.txt:
```diff
@@ -0,0 +1,32 @@
+Developer Guide/Readme for using and extending user settings
```
I will change this to a README.md file, similar to https://github.com/ls1intum/Artemis/tree/develop/src/test/cypress
but for the client-side.
My main goal here is to support further development of the settings for other people and if this guide is directly in the same directory this help will be seen and used more likely than hidden somewhere in the documentation.

Malyuk-A commented on src/main/webapp/app/shared/user-settings/notification-settings/notification-settings.component.html:
```diff
@@ -0,0 +1,43 @@
+<div class="d-flex justify-content-between">
+    <h1>
+        <!--UserSettingsCategory-->
+        {{ 'artemisApp.userSettings.categories.' + (userSettingsCategory | lowerCamelCase) | artemisTranslate }}
```
Good point. I created a comment that refers to the README.md file explaining how the translation works.

Olimbia23 commented on src/main/webapp/app/shared/user-settings/user-settings-guide.txt:
```diff
@@ -0,0 +1,32 @@
+
+    How to create a new "child"-settings.component : (Look at notification-settings as a prime example for child-settings):
+
+        0)  Add a new route for the new settings under userSettingsState (children)
```
step 0?

Malyuk-A replied:
I changed it so the index starts with 1 not 0 anymore

Olimbia23 commented on src/main/webapp/app/shared/user-settings/user-settings-prototype/user-settings-prototype.component.html:
```diff
@@ -0,0 +1,41 @@
+       Exemplary HTML template for UserSettings
+       Every settings page will be somewhat unique, especially its optionCore (different properties and methods)
+       This nested behavior could not be realized via polymorphism due to different types and strong bidirectional dependencies
+       Therefore just copy this template for new settings and modify it. The minimal adjustment needed for the template to work is to implement the
+       option core with its properties and functionalities under <! --OptionCore-- >
+       Look at notification-settings.component.html for a good example
+
+<jhi-alert></jhi-alert>
+<div class="d-flex justify-content-between">
+    <h1>
+        <! --UserSettingsCategory-- >
+        {{ 'artemisApp.userSettings.categories.' + (userSettingsCategory | lowerCamelCase) | artemisTranslate }}
+    </h1>
+    <button type="button" class="btn btn-primary" *ngIf="optionsChanged" id="apply-changes-button" (click)="saveOptions()">
+        <fa-icon [icon]="'save'"></fa-icon>
+        {{ 'artemisApp.userSettings.saveChanges' | artemisTranslate }}
+    </button>
+</div>
+<div class="list-group d-block" *ngIf="userSettings">
+    <div *ngFor="let optionGroup of userSettings.groups">
+        <div *jhiHasAnyAuthority="[optionGroup.restrictionLevel]" class="list-group-item">
+            <h3>
+                <! --GroupName-- >
+                {{ 'artemisApp.userSettings.optionGroupNames.' + (optionGroup.name | lowerCamelCase) | artemisTranslate }}
+            </h3>
+            <div *ngFor="let option of optionGroup.options">
+                <dt>
+                    <! --OptionName-- >
+                    {{ 'artemisApp.userSettings.optionNames.' + (option.name | lowerCamelCase) | artemisTranslate }}
+                </dt>
+                <span>
+                    <! --OptionDescription-- >
+                    {{ 'artemisApp.userSettings.optionDescriptions.' + (option.description | lowerCamelCase) | artemisTranslate }}
+                </span>
+                <! --Put Settings specific core properties & functions here-- >
+            </div>
+        </div>
+    </div>
+</div>
+-->
```
seems like everything is commented out, can we remove it?

Malyuk-A replied:
To support reusability I commented this file out for future (different) settings, so this file can be copied and reused. (Why not just keep the notification-settings template then? -> This notification template may change and differ strongly from other future settings therefore I wanted to provide the bare minimum to be comfortably reused and that supports the new settings data structure hierarchy by dynamic angular template rendering (i.e. ngFor loops)
The problem here also is that the user-settings-prototype.component needs a template. (I tried using "" but this does not work)

Olimbia23 commented on src/test/java/de/tum/in/www1/artemis/NotificationSettingsResourceIntegrationTest.java:
```diff
@@ -0,0 +1,104 @@
+    @Autowired
+    private NotificationOptionRepository notificationOptionRepository;
+
+    private List<User> users;
+
```
Codacy

Olimbia23 reviewed (commented):
Checked the code. Kudos to you for making this huge PR @Malyuk-A 💪
Left some suggestions. Will approve upon addressing.

Malyuk-A commented on src/main/webapp/app/shared/constants/user-settings.constants.ts:
```diff
@@ -0,0 +1,24 @@
+export enum UserSettingsCategory {
+    NOTIFICATION_SETTINGS = 'NOTIFICATION_SETTINGS',
+}
```
The reason is that so far the only user setting available are the ones concerning notifications. But there will definitely be more in the future. Therefore I tried as hard as I can to not only make notification settings but to build a foundation for future developers to add new and modify old settings as easily and reusable as possible.

Malyuk-A replied:
I.e. I created/prepared this enum for further values already. (So future devs don't have to think about reusing already established settings code)

Malyuk-A commented on src/main/webapp/app/shared/pipes/text-to-lower-camel-case.pipe.ts:
```diff
@@ -0,0 +1,27 @@
+import { PipeTransform, Pipe } from '@angular/core';
```
As far as I am aware there is no Angular lowerCamelCase Pipe therefore we need a custom one.
But you are completely correct about the fact that the actual camelCase method called inside the transform() of the pipe should/could be an already established npm implementation. So I switched the custom method inside the transform with the npm camelcase one. 👍

Malyuk-A commented on src/main/webapp/app/shared/notification/notification-sidebar/notification-sidebar.scss:
```diff
@@ -61,4 +58,10 @@
         padding: 20px 16px;
     }
+
+    .header-icon {
+        fa-icon {
+            color: #353d47; /* main navigation bar gray  */
+        }
```
Is fixed

Malyuk-A commented on src/main/webapp/app/shared/layouts/navbar/navbar.component.html:
```diff
@@ -172,6 +172,13 @@
+                    <li>
+                        <a id="settings" class="dropdown-header" [routerLink]="['/user-settings']" (click)="collapseNavbar()">
+                            <fa-icon [icon]="'cog'" [fixedWidth]="true"></fa-icon>
+                            <span jhiTranslate="global.menu.settings">Settings</span>
```
I replaced it with the artemisTranslate pipe now.

Malyuk-A commented on src/main/webapp/app/shared/layouts/navbar/navbar.scss:
```diff
@@ -11,6 +11,17 @@ a:not(.btn):hover {
 }
 
+#settings {
+    color: #353d47; //navigation bar grey
+    font-size: medium;
+    font-weight: 600;
+
+    &:hover {
+        color: #3e8acc; //blue
```
Thank you for the idea and explanation. I reuse global colors now 👍

MerlinMehmed reviewed (approved):
Tested during a testing session on TS3 and it worked as expected.

Olimbia23 reviewed (dismissed):
Thank you for applying my suggestions @Malyuk-A 
Also tested it during testing session. It works nicely. 👍

ndricimrr commented on src/main/webapp/app/shared/user-settings/user-settings.route.ts:
```diff
@@ -0,0 +1,27 @@
+import { AccountInformationComponent } from 'app/shared/user-settings/account-information/account-information.component';
+import { NotificationSettingsComponent } from 'app/shared/user-settings/notification-settings/notification-settings.component';
+
+export const userSettingsState: Routes = [
+    {
+        path: 'user-settings',
+        component: UserSettingsContainerComponent,
+        children: [
+            {
+                // redirects to account information when user-settings are opened
+                path: '',
+                pathMatch: 'full',
+                redirectTo: 'account',
+            },
+            {
+                path: 'account',
+                component: AccountInformationComponent,
+            },
+            {
+                path: 'notifications',
+                component: NotificationSettingsComponent,
+            },
+        ],
+    },
```
I think this is missing route access logic. This might be why trying to access link **/user-settings/notifications** while not being logged in it will not redirect to root route. This file might be helpful: [here](https://github.com/ls1intum/Artemis/blob/bbdfa6c60a010bc276b09909d9a04fe55bc028ef/src/main/webapp/app/exercises/text/manage/tutor-effort/tutor-effort-statistics.route.ts)

Malyuk-A replied:
Great, thank you very much for your help I managed to fix this. 👍

imstreetspirit reviewed (commented):
tested the functionality during testing session on ts3, works as accepted

MaisiKoleni commented on src/main/java/de/tum/in/www1/artemis/web/rest/NotificationResource.java:
```diff
@@ -91,6 +103,30 @@ public NotificationResource(NotificationRepository notificationRepository, UserR
         return new ResponseEntity<>(page.getContent(), headers, HttpStatus.OK);
     }
 
+    /**
+     * GET /notifications/filtered-by-settings : Get all notifications that align with the current user notification settings by pages.
+     *
+     * @param pageable Pagination information for fetching the notifications
+     * @return the filtered list of notifications based on current user settings
+     */
+    @GetMapping("/notifications/filtered-by-settings")
+    @PreAuthorize("hasRole('USER')")
+    public ResponseEntity<List<Notification>> getAllNotificationsForCurrentUserFilteredBySettings(@ApiParam Pageable pageable) {
+        User currentUser = userRepository.getUserWithGroupsAndAuthorities();
+        Set<NotificationOption> notificationOptions = notificationOptionRepository.findAllNotificationOptionsForRecipientWithId(currentUser.getId());
+        Set<NotificationType> deactivatedTypes = notificationSettingsService.findDeactivatedNotificationTypes(notificationOptions);
+        Set<String> deactivatedTitles = notificationSettingsService.convertNotificationTypesToTitles(deactivatedTypes);
+        final Page<Notification> page;
+        if (deactivatedTitles.isEmpty()) {
+            page = notificationRepository.findAllNotificationsForRecipientWithLogin(currentUser.getGroups(), currentUser.getLogin(), pageable);
+        }
+        else {
+            page = notificationRepository.findAllNotificationsFilteredBySettingsForRecipientWithLogin(currentUser.getGroups(), currentUser.getLogin(), deactivatedTitles, pageable);
+        }
+        HttpHeaders headers = PaginationUtil.generatePaginationHttpHeaders(ServletUriComponentsBuilder.fromCurrentRequest(), page);
+        return new ResponseEntity<>(page.getContent(), headers, HttpStatus.OK);
+    }
+
```
So, because it got requested, here is my take on it:

The shown solution has the benefit that we can always retrieve all notifications for the current user or filter them based on the settings.

Now we also have a proposed unified solution (that in this case results in the new method replacing the old, unless I misread something).

The real question to decide this is: what do we need? If the general one is not used anywhere, we can and should replace it. Adding it again in the future is easy, should we need it. If it is used in addition to the new one, we must keep it in some way.

I would, however, also like to provide two new perspectives.

We could enable the filter with an (optional) flag, therefore combining both methods into one. Should work fairly well here because of the similar code and functionality.

One step further would me my preferred way in terms of API design, provided that those options are pretty simple and short. It might however not be that much of an improvement as long as our own client is (pretty much) the only one. This would involve passing the filter options as part of the request, allowing the client to specify what it wants to receive. An empty set for example would just mean _all_ then.
As long as our client is the only supported one and we always filter based on user settings, this would be over-engineered.

So, it depends on what we actually need and use it for. I don't know the details and the context here right now.

One thing however is the REST endpoint. I think, unless we keep both, the endpoint should be called `.../notifications`, because there is no other option and notifications are then always filtered by current user and user settings. Or by additional, optional filter request parameters.

MaisiKoleni replied:
Might also be interesting to know what the rest of the team / Stephan think about that.

MaisiKoleni replied:
One thing to keep in mind in general for such situations, though: the git history will already keep it forever for you.

In the past, I was sometimes reluctant to remove code because it might be useful and used in the future, but in the end it will not be lost anyway and if it stays in the current code base, it was just a burden to work with and might not even get tested. I just deleted it and never had any regrets.

We can be especially brutal here in Artemis because we don't need to care about compatibility (yet).

Malyuk-A replied:
Thank you both very much for your deep dive into this issue. I appreciate your different views and ideas.
After thinking about what you have written I decided to remove the old implementation for receiving all notifications because I could not find a (current) use case where all notifications are used (regardless of settings). And I support your views that it is a good choice to remove things that are not used (we can easily recover/reintroduce it if need be), To leave unused code is an antipattern as we all know (lavaflow).
Thank you 👍

lschlesinger commented on src/main/java/de/tum/in/www1/artemis/web/rest/NotificationSettingsResource.java:
```diff
@@ -0,0 +1,105 @@
+    @PreAuthorize("hasRole('USER')")
+    public ResponseEntity<Set<NotificationOption>> getNotificationOptionsForCurrentUser() {
+        log.debug("REST request to get NotificationOptionsForCurrentUser");
+        long start = System.currentTimeMillis();
```
are you sure we want to record the time for retrieving notification options? Why would I be interested in this information? I am also not sure if we want to log it at all... wouldn't that spam the console?

Malyuk-A replied:
I simply oriented myself based on other resources. I can remove it.

lschlesinger replied:
ah well, I think we do not necessarily need it in this case.

lschlesinger commented on src/main/java/de/tum/in/www1/artemis/web/rest/NotificationSettingsResource.java:
```diff
@@ -0,0 +1,105 @@
+    @PreAuthorize("hasRole('USER')")
+    public ResponseEntity<NotificationOption[]> saveNotificationOptionsForCurrentUser(@RequestBody NotificationOption[] notificationOptions) {
+        log.debug("REST request (POST) to saveUserOptionsForCurrentUser");
+        long start = System.currentTimeMillis();
```
same as above

lschlesinger commented on src/main/resources/config/liquibase/changelog/20210910180000_changelog.xml:
```diff
@@ -0,0 +1,36 @@
+        <addForeignKeyConstraint
+            baseTableName="notification_option"
+            baseColumnNames="user_id"
+            constraintName="fk_user_option_user_id"
```
nit: would fk_notification_option_user_id be more suitable?

Malyuk-A replied:
Yes, it would, this is a big/old PR  problem.  I fear if I change it then e.g. there will be issues on TS3 ...

lschlesinger replied:
then leave it as it is ;)!

lschlesinger commented on src/main/webapp/app/entities/notification.model.ts:
```diff
@@ -22,3 +22,61 @@ export class Notification implements BaseEntity {
 }
+
+/**
+ * Corresponds to the NotificationTitleTypeConstants(.java) constant Strings in the server
```
nit: Corresponds to the server-side NotificationTitleTypeConstants(.java) constant Strings

lschlesinger reviewed (changes requested):
I was not able to review it all... I'll do it in small portions 😄

lschlesinger commented on src/main/webapp/app/shared/user-settings/README.md:
```diff
@@ -0,0 +1,98 @@
+For every new child-settings you have to create a new **table**, **REST controller/resource** (service) and repository due to the possibly big differences between the `OptionCores`.
+            *(Might change due to found similarities)*
+
+
```
why not removing this line to make codacy happy?

lschlesinger commented on src/main/webapp/app/shared/user-settings/README.md:
```diff
@@ -0,0 +1,98 @@
+
+## Additional Information
+For further reading see the original PR that introduces these settings : https://github.com/ls1intum/Artemis/pull/3922
+
```
why not removing this line to make codacy happy?

lschlesinger commented on src/main/webapp/app/shared/user-settings/README.md:
```diff
@@ -0,0 +1,98 @@
+    
+    Add your option to the `user-settings-container.component.html` *(under `<!-- links to different settings -->`)*
+    
+
```
remove line?

lschlesinger commented on src/main/webapp/app/shared/user-settings/README.md:
```diff
@@ -0,0 +1,98 @@
+    
+
+2) **Create a new folder** for `childSettings` and **put** `childSettings` specific files into it
+   
```
remove line?

lschlesinger commented on src/main/webapp/app/shared/user-settings/user-settings-prototype/user-settings-prototype.component.html:
```diff
@@ -0,0 +1,42 @@
+<!--
+       Exemplary HTML template for UserSettings
+       Every settings page will be somewhat unique, especially its optionCore (different properties and methods)
+       This nested behavior could not be realized via polymorphism due to different types and strong bidirectional dependencies
+       Therefore just copy this template for new settings and modify it. The minimal adjustment needed for the template to work is to implement the
+       option core with its properties and functionalities under <! --OptionCore-- >
+       Look at notification-settings.component.html for a good example
+
+<jhi-alert></jhi-alert>
+<div class="d-flex justify-content-between">
+    <h1>
+        <! --UserSettingsCategory-- >
+        <! -- please look at the README.md file to understand how the translation works -- >
+        {{ 'artemisApp.userSettings.categories.' + userSettingsCategory | artemisTranslate }}
+    </h1>
+    <button type="button" class="btn btn-primary" *ngIf="optionsChanged" id="apply-changes-button" (click)="saveOptions()">
+        <fa-icon [icon]="'save'"></fa-icon>
+        {{ 'artemisApp.userSettings.saveChanges' | artemisTranslate }}
+    </button>
+</div>
+<div class="list-group d-block" *ngIf="userSettings">
+    <div *ngFor="let optionGroup of userSettings.groups">
+        <div *jhiHasAnyAuthority="[optionGroup.restrictionLevel]" class="list-group-item">
+            <h3>
+                <! --GroupName-- >
+                {{ 'artemisApp.userSettings.optionGroupNames.' + optionGroup.name | artemisTranslate }}
+            </h3>
+            <div *ngFor="let option of optionGroup.options">
+                <dt>
+                    <! --OptionName-- >
+                    {{ 'artemisApp.userSettings.optionNames.' + option.name | artemisTranslate }}
+                </dt>
+                <span>
+                    <! --OptionDescription-- >
+                    {{ 'artemisApp.userSettings.optionDescriptions.' + option.description | artemisTranslate }}
+                </span>
+                <! --Put Settings specific core properties & functions here-- >
+            </div>
+        </div>
+    </div>
```
I'd argue that this is redundant and also not common to do. With the one instantiation of a user-setting component (`notification-settings.component.ts`) and it's according template (`notification-settings.component.html`) you already provide an example that others can use as orientation. With this "dummy"-component you increase the effort to maintain the code in this folder, as this has to be kept in sync with the `notification-settings.component.html`.

Malyuk-A replied:
Very good points. Yes, I removed the file and refactored the component into a directive thank you very much 🎉

lschlesinger commented on src/main/webapp/app/shared/user-settings/user-settings-prototype/user-settings-prototype.component.ts:
```diff
@@ -0,0 +1,107 @@
+import { OptionCore, UserSettings } from 'app/shared/user-settings/user-settings.model';
+import { UserSettingsService } from 'app/shared/user-settings/user-settings.service';
+
+/**
+ * Is used as the abstract user-settings "parent" component with all the necessary basic logic for other "child" components to implement/inherit from.
+ * The @Component decorated and templateUrl are needed to be able to use OnInit
+ */
+@Component({
+    templateUrl: 'user-settings-prototype.component.html',
```
It's a good idea to provide this base class that concrete instances of the user settings, such as the notification-settings can inherit from.
As far as I know, this is nowadays achieved by using the `@Directive()` decorator, I've read this [here](https://stackoverflow.com/questions/47882247/angular4-components-inheritance-with-abstract-class/47883806#47883806). I also used for metis postings [here](https://github.com/ls1intum/Artemis/blob/develop/src/main/webapp/app/shared/metis/posting.directive.ts)

Malyuk-A replied:
Thank you very much for this information. @Directives are completely new to me and I tried it out and it works like a charm. 🎉

lschlesinger replied:
happy to hear, I also applied it for the first time 😄

Malyuk-A commented on src/main/java/de/tum/in/www1/artemis/domain/notification/GroupNotificationFactory.java:
```diff
@@ -191,13 +188,13 @@ public static GroupNotification createNotification(AnswerPost answerPost, User a
             case NEW_ANSWER_POST_FOR_LECTURE -> {
                 Lecture lecture = answerPost.getPost().getLecture();
-                title = "New Reply";
+                title = NotificationTitleTypeConstants.NEW_POST_FOR_LECTURE_TITLE;
```
Yes, you are correct. Nice catch 👀👍

lschlesinger commented on src/main/webapp/i18n/de/userSettings.json:
```diff
@@ -0,0 +1,46 @@
+                "NOTIFICATION_SETTINGS": "Benachrichtigungseinstellungen"
+            },
+            "optionGroupNames": {
+                "Exercise Notifications": "Aufgabenbenachrichtigungen",
+                "Lecture Notifications": "Vorlesungsbenachrichtigungen",
+                "Instructor Exclusive Notifications": "Exklusive Benachrichtigungen für Lehrende"
```
why not using actual translation keys such as "ExerciseNotifications", "LectureNotifications", "InstructorExclusiveNotifications"?

Malyuk-A replied:
This is related to the way I designed the .defaut.ts settings
![image](https://user-images.githubusercontent.com/65814168/133658468-9f3d2517-4a40-4876-b540-ad2697d6a549.png)
I want to make this and future default settings to be easy to write and read (human-readable not lowerCamelCase)
Because these strings e.g. 'Exercise Notifications' or 'Get notified if a new exercise has been created or started' are easier to type and read than 'getNotifiedIfANewExerciseHasBeenCreatedOrStarted IMO, but if you and others (especially Stephan) tell me to change it to lowerCamelCase I will do so.

That was the entire reason why I played around with a (custom, then imported) new camelCase pipe to keep the default.ts readable and the translation JSON in camelcase format.
(here a picture of the template where the translation takes place) 
![image](https://user-images.githubusercontent.com/65814168/133659197-95849d92-d9ab-490e-8b4b-4b9692fc8709.png)

lschlesinger commented on src/main/webapp/i18n/de/userSettings.json:
```diff
@@ -0,0 +1,46 @@
+                "Instructor Exclusive Notifications": "Exklusive Benachrichtigungen für Lehrende"
+            },
+            "optionNames": {
+                "Exercise created or started": "(Neue) Aufgabe erstellt oder gestartet",
+                "Exercise open for Practice": "Aufgabe zum Üben freigegeben",
+                "New Post for Exercises": "Neue Beiträge für Aufgaben",
+                "New Reply for Exercises": "Neue Antworten für Aufgaben",
+                "Attachment Changes": "Änderungen von Anhängen",
+                "New Post for Lecture": "Neue Beiträge für Vorlesungen",
+                "New Reply for Lecture": "Neue Antworten für Vorlesungen",
+                "Course and Exam Archiving Started": "Kurs und Klausur Archivierungsstart"
```
why not using actual translation keys such as "ExerciseCreatedOrStarted", "ExerciseOpenForPractice"...

Malyuk-A replied:
Please have a look at my response above

lschlesinger commented on src/main/webapp/app/shared/user-settings/notification-settings/notification-settings.default.ts:
```diff
@@ -0,0 +1,104 @@
+            restrictionLevel: Authority.USER,
+            options: [
+                {
+                    name: 'Exercise created or started',
+                    description: 'Get notified if a new exercise has been created or started',
```
why can't we use actual translation keys, such as "ExerciseCreatedOrStarted", I would change this in combination with my comment [here](https://github.com/ls1intum/Artemis/pull/3922/files#r710303615). Or am I missing something completely?

Malyuk-A replied:
Please have a look at my response above

Malyuk-A commented on src/main/java/de/tum/in/www1/artemis/domain/NotificationOption.java:
```diff
@@ -0,0 +1,89 @@
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class NotificationOption extends DomainObject {
+
+    @Column(name = "option_specifier", nullable = false)
+    private String optionSpecifier;
```
There are two reasons why I decided not to use enums on the server:
1) I only use each specifier (by name) once in the mapping (in NotificationSettingsService) so no duplication happens.
2) Based on my experience with enums vs string for the databse. There was a very similar issue with notification types vs title and the question if I could replace the normal string property in the DB for notifications (title) with the enum for notificationTypes but Stephan explained to me that there are issues with the DB and enums if these enums will change often, this does not happen with strings though.

lschlesinger replied:
I see, thanks for clarification! Fine for me!

Malyuk-A commented on src/main/java/de/tum/in/www1/artemis/repository/NotificationOptionRepository.java:
```diff
@@ -0,0 +1,25 @@
+            LEFT JOIN FETCH notificationOption.user user
+            WHERE user.id = :#{#userId}
+            """)
+    Set<NotificationOption> findAllNotificationOptionsForRecipientWithId(@Param("userId") long userId);
```
Thank you for telling me about this feature! I think I already came into contact with it, e.g. notificationOptionRepository.saveAll(...), etc. or is this smth different?
I played around with this now and also tried to find examples but with no luck sadly.

Malyuk-A replied:
I tried it once more but I guess I misunderstand something, but as long as it works.

MerlinMehmed reviewed (commented):
Very good job! I reviewed both client and server code and I left some comments.

Malyuk-A commented on src/main/java/de/tum/in/www1/artemis/web/rest/NotificationSettingsResource.java:
```diff
@@ -0,0 +1,92 @@
+ * REST controller for managing NotificationSettings (NotificationOptions).
+ */
+@RestController
+@RequestMapping("/api")
```
Yes, you are correct thank you for catching this 👍
_(Another thing to keep in mind for the future, but now I know what it is about^^)_

Malyuk-A commented on src/main/java/de/tum/in/www1/artemis/domain/NotificationOption.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Override
+    public int hashCode() {
```
I added the override for equals (as discussed via slack) Thank you very much for your quick replies and help 👍

ndricimrr commented on src/main/java/de/tum/in/www1/artemis/service/NotificationSettingsService.java:
```diff
@@ -0,0 +1,105 @@
+     * @return the converted String Set
+     */
+    public Set<String> convertNotificationTypesToTitles(Set<NotificationType> types) {
+        return types.stream().map(type -> NotificationTitleTypeConstants.findCorrespondingNotificationTitle(type)).collect(Collectors.toSet());
```
```suggestion
        return types.stream().map(NotificationTitleTypeConstants::findCorrespondingNotificationTitle).collect(Collectors.toSet());

```
It seems like this could be replaced with a simpler syntax.

ndricimrr commented on src/main/java/de/tum/in/www1/artemis/service/NotificationSettingsService.java:
```diff
@@ -0,0 +1,105 @@
+     * @return the converted String Set
+     */
+    public Set<String> convertNotificationTypesToTitles(Set<NotificationType> types) {
+        return types.stream().map(type -> NotificationTitleTypeConstants.findCorrespondingNotificationTitle(type)).collect(Collectors.toSet());
```
```suggestion
        return types.stream().map(NotificationTitleTypeConstants::findCorrespondingNotificationTitle).collect(Collectors.toSet());

```
It seems like this could be replaced with a simpler syntax.

Malyuk-A replied:
I added your change to the code thanks  👍

ndricimrr commented on src/main/java/de/tum/in/www1/artemis/service/NotificationSettingsService.java:
```diff
@@ -0,0 +1,105 @@
+     * @param notificationOption which corresponding NotificationTypes should be found
+     * @return the corresponding NotificationType(s)
+     */
+    private NotificationType[] findCorrespondingNotificationTypesForNotificationOption(NotificationOption notificationOption) {
+        switch (notificationOption.getOptionSpecifier()) {
+            case "notification.exercise-notification.exercise-created-or-started": {
+                return new NotificationType[] { NotificationType.EXERCISE_CREATED };
+            }
+            case "notification.exercise-notification.exercise-open-for-practice": {
+                return new NotificationType[] { NotificationType.EXERCISE_PRACTICE };
+            }
+            case "notification.exercise-notification.new-post-exercises": {
+                return new NotificationType[] { NotificationType.NEW_POST_FOR_EXERCISE };
+            }
+            case "notification.exercise-notification.new-answer-post-exercises": {
+                return new NotificationType[] { NotificationType.NEW_ANSWER_POST_FOR_EXERCISE };
+            }
+            case "notification.lecture-notification.attachment-changes": {
+                return new NotificationType[] { NotificationType.ATTACHMENT_CHANGE };
+            }
+            case "notification.lecture-notification.new-post-for-lecture": {
+                return new NotificationType[] { NotificationType.NEW_POST_FOR_LECTURE };
+            }
+            case "notification.lecture-notification.new-answer-post-for-lecture": {
+                return new NotificationType[] { NotificationType.NEW_ANSWER_POST_FOR_LECTURE };
+            }
+            case "notification.instructor-exclusive-notification.course-and-exam-archiving-started": {
+                return new NotificationType[] { NotificationType.EXAM_ARCHIVE_STARTED, NotificationType.COURSE_ARCHIVE_STARTED };
+            }
+            default: {
+                return new NotificationType[0];
+            }
+        }
+    }
```
I took a moment to look into this and it seems like this is causing a high complexity increase **+9** . 
I would suggest to maybe try out this [plugin](https://plugins.jetbrains.com/plugin/12159-codemetrics) to help you locally as I couldn't see the issue on [codacy](https://app.codacy.com/gh/ls1intum/Artemis/file/61236342259/issues/source?fileBranchId=25004975).

I looked into it a bit, and I think there's a much simpler way of doing this, with less code and complexity. Psuedocode below:

```
    Map<String, NotificationType[]> mappings = Map.of(
        "notification.exercise-notification.exercise-created-or-started", new NotificationType[]{EXERCISE_CREATED},
        "notification.exercise-notification.exercise-open-for-practice", new NotificationType[]{EXERCISE_PRACTICE},
        "notification.exercise-notification.new-post-exercises", new NotificationType[]{NEW_POST_FOR_EXERCISE},
        "notification.exercise-notification.new-answer-post-exercises", new NotificationType[]{NEW_ANSWER_POST_FOR_EXERCISE},
        "notification.lecture-notification.attachment-changes", new NotificationType[]{ATTACHMENT_CHANGE },
        "notification.lecture-notification.new-post-for-lecture", new NotificationType[]{NEW_POST_FOR_LECTURE }
    );

    private NotificationType[] findCorrespondingNotificationTypesForNotificationOption(NotificationOption notificationOption) {
        return mappings.getOrDefault(notificationOption.getOptionSpecifier(), new NotificationType[0]);
    }
```

So maybe have a predefined set of mappings, and then use map get. You can avoid,  NotificationType.NEW_ANSWER_POST_FOR_LECTURE and use NEW_ANSWER_POST_FOR_LECTURE  directly if you import it statically.

Malyuk-A replied:
You are 100% correct. I now refactored this and the corresponding part on the client-side by using maps.
Now the code is cleaner and complexity lower. Thank you 👍

ndricimrr reviewed (commented):
I looked again and added some code suggestions. Might make it easier to read, and less complexity.

ndricimrr commented on src/main/java/de/tum/in/www1/artemis/web/rest/NotificationSettingsResource.java:
```diff
@@ -0,0 +1,92 @@
+     */
+    @PostMapping("notification-settings/save-options")
+    @PreAuthorize("hasRole('USER')")
+    public ResponseEntity<NotificationOption[]> saveNotificationOptionsForCurrentUser(@RequestBody NotificationOption[] notificationOptions) {
+        if (notificationOptions == null || notificationOptions.length == 0) {
```
```suggestion
    public ResponseEntity<NotificationOption[]> saveNotificationOptionsForCurrentUser(@NotNull @RequestBody NotificationOption[] notificationOptions) {
        if (notificationOptions.length == 0) {
```
this method seems to have a pretty high complexity as well. Maybe we could somehow incorporate `@NotNull `and void the null check. Sounds [doable](https://stackoverflow.com/a/57038385), but I haven't tested this. What do you think?

ndricimrr commented on src/main/webapp/app/shared/user-settings/notification-settings/notification-settings.service.ts:
```diff
@@ -0,0 +1,91 @@
+     * @return the corresponding NotificationType(s)
+     */
+    private static findCorrespondingNotificationTypesForNotificationOptionCore(notificationOptionCore: NotificationOptionCore): string[] {
+        switch (notificationOptionCore.optionSpecifier) {
+            case OptionSpecifier.NOTIFICATION__EXERCISE_NOTIFICATION__EXERCISE_CREATED_OR_STARTED: {
+                return [EXERCISE_CREATED_TITLE];
+            }
+            case OptionSpecifier.NOTIFICATION__EXERCISE_NOTIFICATION__EXERCISE_OPEN_FOR_PRACTICE: {
+                return [EXERCISE_PRACTICE_TITLE];
+            }
+            case OptionSpecifier.NOTIFICATION__EXERCISE_NOTIFICATION__NEW_POST_EXERCISES: {
+                return [NEW_POST_FOR_EXERCISE_TITLE];
+            }
+            case OptionSpecifier.NOTIFICATION__EXERCISE_NOTIFICATION__NEW_ANSWER_POST_EXERCISES: {
+                return [NEW_ANSWER_POST_FOR_EXERCISE_TITLE];
+            }
+            case OptionSpecifier.NOTIFICATION__LECTURE_NOTIFICATION__ATTACHMENT_CHANGES: {
+                return [ATTACHMENT_CHANGE_TITLE];
+            }
+            case OptionSpecifier.NOTIFICATION__LECTURE_NOTIFICATION__NEW_POST_FOR_LECTURE: {
+                return [NEW_POST_FOR_LECTURE_TITLE];
+            }
+            case OptionSpecifier.NOTIFICATION__LECTURE_NOTIFICATION__NEW_ANSWER_POST_FOR_LECTURE: {
+                return [NEW_ANSWER_POST_FOR_LECTURE_TITLE];
+            }
+            case OptionSpecifier.NOTIFICATION__INSTRUCTOR_EXCLUSIVE_NOTIFICATIONS__COURSE_AND_EXAM_ARCHIVING_STARTED: {
+                return [EXAM_ARCHIVE_STARTED_TITLE, COURSE_ARCHIVE_STARTED_TITLE];
+            }
+            default: {
+                return [];
+            }
+        }
+    }
```
This file also seems to have a high complexity. It seems similiar to the java counterpart you had. 
Maybe you could also refactor this with much less code and complexity. I would suggest something like this: (pseudocode, not tested):

```
const mappings: Map<OptionSpecifier, string[]> = new Map([
            [OptionSpecifier.NOTIFICATION__EXERCISE_NOTIFICATION__EXERCISE_CREATED_OR_STARTED, [EXERCISE_CREATED_TITLE]],
            [OptionSpecifier.NOTIFICATION__EXERCISE_NOTIFICATION__EXERCISE_OPEN_FOR_PRACTICE, [EXERCISE_PRACTICE_TITLE]],
            [OptionSpecifier.NOTIFICATION__EXERCISE_NOTIFICATION__NEW_POST_EXERCISES, [NEW_POST_FOR_EXERCISE_TITLE]],
            [OptionSpecifier.NOTIFICATION__EXERCISE_NOTIFICATION__NEW_ANSWER_POST_EXERCISES, [NEW_ANSWER_POST_FOR_EXERCISE_TITLE]],
            [OptionSpecifier.NOTIFICATION__LECTURE_NOTIFICATION__ATTACHMENT_CHANGES, [ATTACHMENT_CHANGE_TITLE]],
            [OptionSpecifier.NOTIFICATION__LECTURE_NOTIFICATION__NEW_POST_FOR_LECTURE, [NEW_POST_FOR_LECTURE_TITLE]],
            [OptionSpecifier.NOTIFICATION__LECTURE_NOTIFICATION__NEW_ANSWER_POST_FOR_LECTURE, [NEW_ANSWER_POST_FOR_LECTURE_TITLE]],
            [OptionSpecifier.NOTIFICATION__INSTRUCTOR_EXCLUSIVE_NOTIFICATIONS__COURSE_AND_EXAM_ARCHIVING_STARTED, [EXAM_ARCHIVE_STARTED_TITLE, COURSE_ARCHIVE_STARTED_TITLE]]
        ]);
        const foundOption = mappings.get(notificationOptionCore.optionSpecifier);
        return foundOption ? foundOption : [];
```
I think this is much neater, and does the same in less lines.

ndricimrr reviewed (commented):
Found some more places for further improvements. Let me know if it makes sense.

Malyuk-A commented on src/main/java/de/tum/in/www1/artemis/web/rest/NotificationSettingsResource.java:
```diff
@@ -0,0 +1,92 @@
+        log.debug("REST request to save NotificationOptions : {} for current user {}", notificationOptions, currentUser);
+        notificationSettingsService.setCurrentUser(notificationOptions, currentUser);
+        List<NotificationOption> resultAsList = notificationOptionRepository.saveAll(Arrays.stream(notificationOptions).toList());
+        if (resultAsList == null || resultAsList.isEmpty()) {
```
Yes, you are right.

ndricimrr commented on src/main/java/de/tum/in/www1/artemis/service/NotificationSettingsService.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    private final static Map<String, NotificationType[]> NOTIFICATION_OPTION_SPECIFIER_TO_NOTIFICATION_TYPES_MAP = Map.of(
+            "notification.exercise-notification.exercise-created-or-started", new NotificationType[] { EXERCISE_CREATED },
+            "notification.exercise-notification.exercise-open-for-practice", new NotificationType[] { EXERCISE_PRACTICE }, "notification.exercise-notification.new-post-exercises",
+            new NotificationType[] { NEW_POST_FOR_EXERCISE }, "notification.exercise-notification.new-answer-post-exercises",
+            new NotificationType[] { NEW_ANSWER_POST_FOR_EXERCISE }, "notification.lecture-notification.attachment-changes", new NotificationType[] { ATTACHMENT_CHANGE },
```
[nit] this doesn't look very prettified 😄 . or does the commit hook automatically keep the line long like this. maybe keep them one entry per line, looks more readable.

Malyuk-A replied:
I see, smth happened during commiting/pushing ...

Malyuk-A replied:
I edited it in gitHub directly so now it should be fine^^

Malyuk-A replied:
This lead to codacy issues so I looked at the code locally again, made the style as it should (each line one entry) and even executen prettier on it but it again changed it during the commiting/pushing to this ... strange

Malyuk-A replied:
I have now changed the initialization of the map. This should do the trick

Olimbia23 reviewed (dismissed):
Reapprove 👍

ndricimrr commented on src/main/java/de/tum/in/www1/artemis/service/NotificationSettingsService.java:
```diff
@@ -0,0 +1,85 @@
+    private Map<NotificationType, Boolean> convertNotificationOptionsToNotificationTypesWithActivationStatus(Set<NotificationOption> notificationOptions) {
+        Map<NotificationType, Boolean> resultingMap = new HashMap<>();
+        for (NotificationOption option : notificationOptions) {
+            NotificationType[] tmpNotificationTypes = NOTIFICATION_OPTION_SPECIFIER_TO_NOTIFICATION_TYPES_MAP.get(option.getOptionSpecifier());
+            if (tmpNotificationTypes != null) {
+                for (NotificationType type : tmpNotificationTypes) {
+                    resultingMap.put(type, option.isWebapp());
+                }
+            }
```
```suggestion
            NotificationType[] tmpNotificationTypes = NOTIFICATION_OPTION_SPECIFIER_TO_NOTIFICATION_TYPES_MAP.getOrDefault(option.getOptionSpecifier(), new NotificationType[0]);
            for (NotificationType type : tmpNotificationTypes) {
                resultingMap.put(type, option.isWebapp());
            }
```

I think this could also be improved. The for loop won't loop if list is empty, no need for null check.

ndricimrr reviewed (approved):
Thanks for applying suggestions. Looks good 👍

Olimbia23 reviewed (dismissed):
Reapprove 👍

MerlinMehmed reviewed (dismissed):
Changes were addressed and the functionality still works as expected. Reapprove

DanielKainz reviewed (approved):
I tested it on TS1. Everything works as intended. Great work 🎉

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/NotificationSettingsResource.java:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @return the list of found NotificationOptions
+     */
+    @GetMapping("notification-settings/fetch-options")
```
This is not the correct REST URL, see the guidelines.

krusche replied:
```suggestion
    @GetMapping("notification-settings")
```

Malyuk-A replied:
I applied the change.

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/NotificationSettingsResource.java:
```diff
@@ -0,0 +1,94 @@
+     * @return the UserOptions that just got saved for the current user as array
+     * 200 for a successful execution, 400 if the user provided empty options to save, 500 if the save call returns empty options
+     */
+    @PostMapping("notification-settings/save-options")
```
same here

krusche replied:
```suggestion
    @PostMapping("notification-settings")
```

Malyuk-A replied:
I applied the change.

Malyuk-A commented:
I implemented the change request from @krusche. I.e. I renamed the option to setting (server and client side), moved the default settings to the server (default-settings in the client now only consist of permanent hierarchical structures and translations, therefore I renamed them to settings-structures). The REST loading call now always returns settings.

I tested it locally, everything works as intended (so far the test servers are occupied).
The failing TypeScript Client Test confuses me because all client tests work for me locally and the error in the logs do not really make sense to me (issue seems to be based on used dependencies, but mine are the exact same as the newest ones)
![image](https://user-images.githubusercontent.com/65814168/134244748-6a8a92aa-1b93-4ee9-ab99-19836b9801ee.png)

lschlesinger reviewed (commented):
Regarding the code, I only have some nitpicks, which do not affect functionality. But I'd change the styling of the settings option in the dropdown to be more consistent with the other options there, i.e. similar to Sign out. I add a picture that shows I would prefer the look and what you have to change
<img width="457" alt="Bildschirmfoto 2021-09-28 um 13 46 44" src="https://user-images.githubusercontent.com/41366522/135092301-a797e1a7-22a3-40d4-bd16-2a37b01f1ec5.png">

lschlesinger reviewed (approved):
I approve this PR after having checked it for several times 👍

lschlesinger reviewed (approved):
re-approve after font size change

ndricimrr reviewed (approved):
I just tested again on TS2 and it worked well as described 👍 
Re-Approve

DanielKainz reviewed (approved):
I tested it again on TS1. Everything works as expected. Great work :-)
Re-approve

Malyuk-A commented:
Quick update: Because this PR takes quite a long time to be merged, naturally a merge conflict occurred.
The change is very very small. I only replaced the name of a constant: 
![image](https://user-images.githubusercontent.com/65814168/135753234-b573bcbc-83da-46ff-8ad3-a010d07ede81.png)

All tests still work : 
![image](https://user-images.githubusercontent.com/65814168/135753240-999acb92-9bb5-49f7-8be7-92dae5aea064.png)

The PR has been extensively tested multiple times and approved: 
![image](https://user-images.githubusercontent.com/65814168/135753266-9a8f636c-4c2d-4f60-8b4b-c6093f09e6ef.png)

=> It is still mergable

Olimbia23 reviewed (approved):
Reapprove 👍

lschlesinger reviewed (approved):
I checked the non-critical, minor change in the last commit. Re-Approve.

ndricimrr reviewed (approved):
Looks good. Only minor merge-conflict changed. Re-Approve. 👍

krusche commented:
@Malyuk-A As this is a large PR, I would appreciate some more reviews and tests.

Malyuk-A commented:
Thank you all very much for testing, your feedback!
The requested changes will be fixed in the email support or a smaller follow-up PR. _(The mentioned sidebar issues are legacy-related because we could observe similar behavior on development.)_
So I think It is ok to merge this PR.
